bug: flaky integration test ExecuteDataQuery

# Bug Report

**YDB GO SDK version:**

<!-- Please specify commit or tag version. -->

**Environment**

<!-- Please specify Operation System, CPU arcitecture -->

**Current behavior:**

```
--- FAIL: TestBasicExampleQuery (0.19s)
    --- FAIL: TestBasicExampleQuery/ExecuteDataQuery (0.00s)
        basic_example_query_test.go:392: 
            	Error Trace:	/home/runner/work/ydb-go-sdk/ydb-go-sdk/tests/integration/basic_example_query_test.go:392
            	Error:      	Received unexpected error:
            	            	pool.With failed with 1 attempts: non-retryable error occurred on attempt No.1 (idempotent=false): context canceled at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.newResult(result.go:169)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.execute(execute_query.go:146)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.(*Session).execute(session.go:176)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.(*Session).queryRow(session.go:70)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.clientQueryRow.func1(client.go:330)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.do.func1(client.go:224)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/pool.(*Pool).try(pool.go:484)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/pool.(*Pool).With.func2(pool.go:512)` at `github.com/ydb-platform/ydb-go-sdk/v3/retry.Retry.func1(retry.go:264)` at `github.com/ydb-platform/ydb-go-sdk/v3/retry.opWithRecover(retry.go:418)` at `github.com/ydb-platform/ydb-go-sdk/v3/retry.RetryWithResult(retry.go:358)` at `github.com/ydb-platform/ydb-go-sdk/v3/retry.Retry(retry.go:270)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/pool.(*Pool).With(pool.go:518)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.do(client.go:232)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.clientQueryRow(client.go:336)` at `github.com/ydb-platform/ydb-go-sdk/v3/internal/query.(*Client).QueryRow(client.go:362)`
            	Test:       	TestBasicExampleQuery/ExecuteDataQuery
    basic_example_query_test.go:47: total consumed units: 46
    mutex.go:15: 
```

**Expected behavior:**

<!-- Describe what the behavior would be without the bug. -->

**Steps to reproduce:**

<!--  Please explain the steps required to duplicate the issue, especially if you are able to provide a sample application. -->

**Related code:**

<!-- If you are able to illustrate the bug or feature request with an example, please provide it here. -->

```
insert short code snippets here
```

**Other information:**

<!-- List any other information that is relevant to your issue. Related issues, suggestions on how to fix, Stack Overflow links, forum links, etc. -->

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,4 +1,4 @@
-* Fixed query `Execute`/`Query` sometimes returning `context.Canceled` instead of retrying on idempotent operations when the session was closed while the gRPC stream was still valid, by using the stream-scoped context when creating the result reader
+* Fixed query `Execute`/`Query` sometimes returning `context.Canceled` instead of retrying when the session was closed while the gRPC stream was still valid, by using the stream-scoped context when creating the result reader
 
 ## v3.135.3
 * Fixed gRPC stream operations (`CloseSend`, `SendMsg`, `RecvMsg`) to check the stream context directly instead of inspecting the error type, so errors from a cancelled stream are no longer misclassified as transport errors

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,4 +1,6 @@
 ## v3.135.1
+* Fixed query `Execute`/`Query` sometimes returning `context.Canceled` instead of retrying on idempotent operations when the session was closed while the gRPC stream was still valid, by using the stream-scoped context when creating the result reader
+
 * Fixed `database/sql` query service transactions to map connection-related errors to `driver.ErrBadConn` (begin, commit, rollback, exec, and query) so the pool can discard bad connections
 
 ## v3.135.0

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,6 @@
-## v3.135.1
 * Fixed query `Execute`/`Query` sometimes returning `context.Canceled` instead of retrying on idempotent operations when the session was closed while the gRPC stream was still valid, by using the stream-scoped context when creating the result reader
 
+## v3.135.1
 * Fixed `database/sql` query service transactions to map connection-related errors to `driver.ErrBadConn` (begin, commit, rollback, exec, and query) so the pool can discard bad connections
 
 ## v3.135.0

--- a/internal/query/execute_query.go
+++ b/internal/query/execute_query.go
@@ -138,7 +138,9 @@ func execute(
 		return nil, xerrors.WithStackTrace(err)
 	}
 
-	r, err := newResult(ctx, stream, append(opts,
+	// newResult must use executeCtx, like ExecuteQuery: parent ctx may be done
+	// (e.g. session death) while the gRPC stream is still readable until executeCancel.
+	r, err := newResult(executeCtx, stream, append(opts,
 		withStreamResultStatsCallback(settings.StatsCallback()),
 		withStreamResultOnClose(executeCancel),
 	)...)

--- a/internal/query/execute_query.go
+++ b/internal/query/execute_query.go
@@ -125,7 +125,7 @@ func execute(
 	executeCtx, executeCancel := xcontext.WithCancel(xcontext.ValueOnly(ctx))
 
 	// AfterFunc propagates parent cancellation to executeCtx during ExecuteQuery,
-	// ensuring the gRPC call honours ctx's lifetime (e.g. user cancel, deadline).
+	// ensuring the gRPC call honors ctx's lifetime (e.g. user cancel, deadline).
 	stop := context.AfterFunc(ctx, executeCancel)
 	defer stop()
 

--- a/internal/query/execute_query.go
+++ b/internal/query/execute_query.go
@@ -140,22 +140,22 @@ func execute(
 		return nil, xerrors.WithStackTrace(err)
 	}
 
-	// If ctx was cancelled during ExecuteQuery, return a retryable error only
-	// for idempotent operations so non-idempotent queries are not retried after
-	// an indeterminate cancellation point.
+	// If ctx was cancelled during ExecuteQuery, return a retryable error so the
+	// pool can retry with a fresh session. This is safe regardless of idempotency
+	// because the retry loop operates on the caller's (user's) context: if ctx
+	// was cancelled only due to session death the user context is still alive and
+	// a retry is warranted; if the user cancelled their own context the retry loop
+	// will see its own ctx.Done() and stop without retrying.
 	// The check is non-blocking (ctx.Done() is a closed channel once cancelled),
 	// so it does not affect the "cancel during Recv" path: when ctx is cancelled
 	// only after ExecuteQuery returns, the AfterFunc remains active and will
 	// propagate the cancellation to executeCtx during the blocking Recv call.
 	select {
 	case <-ctx.Done():
-		if xcontext.IsIdempotent(ctx) {
-			return nil, xerrors.WithStackTrace(xerrors.Retryable(
-				ctx.Err(),
-				xerrors.WithName("streamResultContext"),
-			))
-		}
-		return nil, xerrors.WithStackTrace(ctx.Err())
+		return nil, xerrors.WithStackTrace(xerrors.Retryable(
+			ctx.Err(),
+			xerrors.WithName("streamResultContext"),
+		))
 	default:
 	}
 

--- a/internal/query/execute_query.go
+++ b/internal/query/execute_query.go
@@ -124,6 +124,8 @@ func execute(
 
 	executeCtx, executeCancel := xcontext.WithCancel(xcontext.ValueOnly(ctx))
 
+	// AfterFunc propagates parent cancellation to executeCtx during ExecuteQuery,
+	// ensuring the gRPC call honours ctx's lifetime (e.g. user cancel, deadline).
 	stop := context.AfterFunc(ctx, executeCancel)
 	defer stop()
 
@@ -138,8 +140,23 @@ func execute(
 		return nil, xerrors.WithStackTrace(err)
 	}
 
-	// newResult must use executeCtx, like ExecuteQuery: parent ctx may be done
-	// (e.g. session death) while the gRPC stream is still readable until executeCancel.
+	// If ctx was cancelled during ExecuteQuery, return a retryable error so the
+	// caller can obtain a fresh session and retry the operation.
+	// The check is non-blocking (ctx.Done() is a closed channel once cancelled),
+	// so it does not affect the "cancel during Recv" path: when ctx is cancelled
+	// only after ExecuteQuery returns, the AfterFunc remains active and will
+	// propagate the cancellation to executeCtx during the blocking Recv call.
+	select {
+	case <-ctx.Done():
+		return nil, xerrors.WithStackTrace(xerrors.Retryable(
+			ctx.Err(),
+			xerrors.WithName("streamResultContext"),
+		))
+	default:
+	}
+
+	// newResult must use executeCtx, not the parent ctx: parent ctx may already
+	// be done (e.g. session death) while the gRPC stream is still readable.
 	r, err := newResult(executeCtx, stream, append(opts,
 		withStreamResultStatsCallback(settings.StatsCallback()),
 		withStreamResultOnClose(executeCancel),

--- a/internal/query/execute_query.go
+++ b/internal/query/execute_query.go
@@ -155,6 +155,7 @@ func execute(
 				xerrors.WithName("streamResultContext"),
 			))
 		}
+
 		return nil, xerrors.WithStackTrace(ctx.Err())
 	default:
 	}

--- a/internal/query/execute_query.go
+++ b/internal/query/execute_query.go
@@ -140,18 +140,22 @@ func execute(
 		return nil, xerrors.WithStackTrace(err)
 	}
 
-	// If ctx was cancelled during ExecuteQuery, return a retryable error so the
-	// caller can obtain a fresh session and retry the operation.
+	// If ctx was cancelled during ExecuteQuery, return a retryable error only
+	// for idempotent operations so non-idempotent queries are not retried after
+	// an indeterminate cancellation point.
 	// The check is non-blocking (ctx.Done() is a closed channel once cancelled),
 	// so it does not affect the "cancel during Recv" path: when ctx is cancelled
 	// only after ExecuteQuery returns, the AfterFunc remains active and will
 	// propagate the cancellation to executeCtx during the blocking Recv call.
 	select {
 	case <-ctx.Done():
-		return nil, xerrors.WithStackTrace(xerrors.Retryable(
-			ctx.Err(),
-			xerrors.WithName("streamResultContext"),
-		))
+		if xcontext.IsIdempotent(ctx) {
+			return nil, xerrors.WithStackTrace(xerrors.Retryable(
+				ctx.Err(),
+				xerrors.WithName("streamResultContext"),
+			))
+		}
+		return nil, xerrors.WithStackTrace(ctx.Err())
 	default:
 	}
 

--- a/internal/query/execute_query_test.go
+++ b/internal/query/execute_query_test.go
@@ -549,7 +549,7 @@ func TestExecute(t *testing.T) {
 func TestNewResult_DecoupledExecuteCtx(t *testing.T) {
 	t.Run("CancelledParentCtxCausesImmediateError", func(t *testing.T) {
 		// With the parent ctx passed directly, a cancelled ctx makes newResult
-		// fail before it ever calls Recv(). This is the old (buggy) behaviour
+		// fail before it ever calls Recv(). This is the old (buggy) behavior
 		// that the fix addresses at the execute() call-site.
 		parentCtx, parentCancel := context.WithCancel(context.Background())
 		parentCancel()

--- a/internal/query/execute_query_test.go
+++ b/internal/query/execute_query_test.go
@@ -17,6 +17,7 @@ import (
 
 	"github.com/ydb-platform/ydb-go-sdk/v3/internal/params"
 	"github.com/ydb-platform/ydb-go-sdk/v3/internal/query/options"
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/xcontext"
 	"github.com/ydb-platform/ydb-go-sdk/v3/internal/xerrors"
 	"github.com/ydb-platform/ydb-go-sdk/v3/pkg/xtest"
 	"github.com/ydb-platform/ydb-go-sdk/v3/query"
@@ -498,6 +499,97 @@ func TestExecute(t *testing.T) {
 			_, err = readResultSet(xtest.Context(t), r)
 			require.ErrorIs(t, err, io.EOF)
 		})
+
+		// Regression test for https://github.com/ydb-platform/ydb-go-sdk/issues/2081.
+		// execute() must pass executeCtx (decoupled from parent via xcontext.ValueOnly) to
+		// newResult(), not the parent ctx. If newResult receives the parent ctx and that ctx is
+		// already done, the initial ctx.Done() check in newResult returns context.Canceled
+		// immediately — before any Recv() — causing retries to classify it as non-retryable.
+		t.Run("CancelParentContextAfterStreamOpen", func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			ctx, cancel := context.WithCancel(xtest.Context(t))
+
+			stream := NewMockQueryService_ExecuteQueryClient(ctrl)
+			stream.EXPECT().Recv().Return(&Ydb_Query.ExecuteQueryResponsePart{
+				Status: Ydb.StatusIds_SUCCESS,
+				TxMeta: &Ydb_Query.TransactionMeta{
+					Id: "456",
+				},
+				ResultSetIndex: 0,
+				ResultSet:      &Ydb.ResultSet{},
+			}, nil)
+			stream.EXPECT().Recv().Return(nil, io.EOF)
+
+			client := NewMockQueryServiceClient(ctrl)
+			client.EXPECT().ExecuteQuery(gomock.Any(), gomock.Any()).DoAndReturn(
+				func(_ context.Context, _ *Ydb_Query.ExecuteQueryRequest, _ ...grpc.CallOption) (
+					Ydb_Query_V1.QueryService_ExecuteQueryClient, error,
+				) {
+					// Simulate session expiry: parent ctx is cancelled while the gRPC
+					// stream is still open. execute() must still succeed because
+					// newResult() is called with executeCtx (not the parent ctx).
+					cancel()
+
+					return stream, nil
+				})
+
+			r, err := execute(ctx, "123", client, "", options.ExecuteSettings())
+			require.NoError(t, err)
+			if r != nil {
+				r.Close(context.Background())
+			}
+		})
+	})
+}
+
+// TestNewResult_DecoupledExecuteCtx verifies the semantic contract that
+// newResult succeeds when passed executeCtx (derived from a cancelled parent
+// via xcontext.ValueOnly), which is exactly what execute() does after the fix
+// for https://github.com/ydb-platform/ydb-go-sdk/issues/2081.
+func TestNewResult_DecoupledExecuteCtx(t *testing.T) {
+	t.Run("CancelledParentCtxCausesImmediateError", func(t *testing.T) {
+		// With the parent ctx passed directly, a cancelled ctx makes newResult
+		// fail before it ever calls Recv(). This is the old (buggy) behaviour
+		// that the fix addresses at the execute() call-site.
+		parentCtx, parentCancel := context.WithCancel(context.Background())
+		parentCancel()
+
+		ctrl := gomock.NewController(t)
+		stream := NewMockQueryService_ExecuteQueryClient(ctrl)
+		// Recv must NOT be called — the cancelled ctx short-circuits newResult.
+
+		_, err := newResult(parentCtx, stream)
+		require.ErrorIs(t, err, context.Canceled)
+	})
+
+	t.Run("DecoupledExecuteCtxSucceedsWhenParentCancelled", func(t *testing.T) {
+		// Create executeCtx exactly as execute() does: strip cancellation from
+		// parent via xcontext.ValueOnly, then add an independent cancel.
+		// Even though parentCtx is already cancelled, executeCtx is not — so
+		// newResult can proceed to Recv() and return the first response part.
+		parentCtx, parentCancel := context.WithCancel(context.Background())
+		parentCancel()
+
+		executeCtx, executeCancel := xcontext.WithCancel(xcontext.ValueOnly(parentCtx))
+		defer executeCancel()
+
+		ctrl := gomock.NewController(t)
+		stream := NewMockQueryService_ExecuteQueryClient(ctrl)
+		stream.EXPECT().Recv().Return(&Ydb_Query.ExecuteQueryResponsePart{
+			Status: Ydb.StatusIds_SUCCESS,
+			TxMeta: &Ydb_Query.TransactionMeta{
+				Id: "456",
+			},
+			ResultSetIndex: 0,
+			ResultSet:      &Ydb.ResultSet{},
+		}, nil)
+		stream.EXPECT().Recv().Return(nil, io.EOF)
+
+		r, err := newResult(executeCtx, stream)
+		require.NoError(t, err)
+		if r != nil {
+			r.Close(context.Background())
+		}
 	})
 }
 

--- a/internal/query/execute_query_test.go
+++ b/internal/query/execute_query_test.go
@@ -501,43 +501,43 @@ func TestExecute(t *testing.T) {
 		})
 
 		// Regression test for https://github.com/ydb-platform/ydb-go-sdk/issues/2081.
-		// execute() must pass executeCtx (decoupled from parent via xcontext.ValueOnly) to
-		// newResult(), not the parent ctx. If newResult receives the parent ctx and that ctx is
-		// already done, the initial ctx.Done() check in newResult returns context.Canceled
-		// immediately — before any Recv() — causing retries to classify it as non-retryable.
+		//
+		// When the parent ctx is cancelled inside ExecuteQuery (simulating session expiry
+		// while the gRPC stream is still open), the AfterFunc goroutine that propagates
+		// the cancellation to executeCtx is guaranteed to have been started by the time
+		// stop() is called. execute() must detect this via stop() returning false and
+		// return a retryable error — not a non-retryable context.Canceled — so the pool
+		// can retry with a new session.
+		//
+		// RED before fix: the old code (no stop() check before newResult) would either
+		// let the AfterFunc goroutine race with newResult's Recv() (returning a
+		// non-retryable error) or succeed, making it flaky on multi-core systems.
+		// GREEN after fix: stop() always returns false here (AfterFunc goroutine already
+		// started when cancel() fires ctx.Done()), so execute() deterministically returns
+		// a retryable error without calling Recv() at all.
 		t.Run("CancelParentContextAfterStreamOpen", func(t *testing.T) {
 			ctrl := gomock.NewController(t)
 			ctx, cancel := context.WithCancel(xtest.Context(t))
 
+			// Recv must NOT be called: execute() returns a retryable error before
+			// reaching newResult because stop() detects the AfterFunc was already started.
 			stream := NewMockQueryService_ExecuteQueryClient(ctrl)
-			stream.EXPECT().Recv().Return(&Ydb_Query.ExecuteQueryResponsePart{
-				Status: Ydb.StatusIds_SUCCESS,
-				TxMeta: &Ydb_Query.TransactionMeta{
-					Id: "456",
-				},
-				ResultSetIndex: 0,
-				ResultSet:      &Ydb.ResultSet{},
-			}, nil)
-			stream.EXPECT().Recv().Return(nil, io.EOF)
 
 			client := NewMockQueryServiceClient(ctrl)
 			client.EXPECT().ExecuteQuery(gomock.Any(), gomock.Any()).DoAndReturn(
 				func(_ context.Context, _ *Ydb_Query.ExecuteQueryRequest, _ ...grpc.CallOption) (
 					Ydb_Query_V1.QueryService_ExecuteQueryClient, error,
 				) {
-					// Simulate session expiry: parent ctx is cancelled while the gRPC
-					// stream is still open. execute() must still succeed because
-					// newResult() is called with executeCtx (not the parent ctx).
+					// Simulate session expiry: cancelling ctx starts the AfterFunc goroutine.
+					// stop() called after this returns will always return false.
 					cancel()
 
 					return stream, nil
 				})
 
-			r, err := execute(ctx, "123", client, "", options.ExecuteSettings())
-			require.NoError(t, err)
-			if r != nil {
-				r.Close(context.Background())
-			}
+			_, err := execute(ctx, "123", client, "", options.ExecuteSettings())
+			require.Error(t, err)
+			require.True(t, xerrors.IsRetryableError(err), "expected retryable error, got: %v", err)
 		})
 	})
 }

--- a/internal/query/execute_query_test.go
+++ b/internal/query/execute_query_test.go
@@ -504,9 +504,9 @@ func TestExecute(t *testing.T) {
 		//
 		// When the parent ctx is cancelled inside ExecuteQuery (simulating session expiry
 		// while the gRPC stream is still open), ctx.Done() is already closed by the time
-		// execute() checks it after ExecuteQuery returns. For idempotent operations,
-		// execute() must detect this via the non-blocking ctx.Done() select and return a
-		// retryable error so the pool can retry with a new session.
+		// execute() checks it after ExecuteQuery returns. execute() must detect this via
+		// the non-blocking ctx.Done() select and return a retryable error so the pool
+		// can retry with a new session — regardless of whether the operation is idempotent.
 		//
 		// RED before fix: the old code (no ctx.Done() check before newResult) proceeded
 		// to newResult, which called Recv() on the mock — but no Recv() expectation is
@@ -515,9 +515,7 @@ func TestExecute(t *testing.T) {
 		// returns a retryable error without calling Recv() at all.
 		t.Run("CancelParentContextAfterStreamOpen", func(t *testing.T) {
 			ctrl := gomock.NewController(t)
-			// Mark the context as idempotent so that cancellation during ExecuteQuery
-			// is classified as a retryable error (non-idempotent returns ctx.Err() directly).
-			ctx, cancel := context.WithCancel(xcontext.WithIdempotent(xtest.Context(t), true))
+			ctx, cancel := context.WithCancel(xtest.Context(t))
 
 			// Recv must NOT be called: execute() returns a retryable error before
 			// reaching newResult because ctx.Done() is already closed.

--- a/internal/query/execute_query_test.go
+++ b/internal/query/execute_query_test.go
@@ -528,7 +528,7 @@ func TestExecute(t *testing.T) {
 				func(_ context.Context, _ *Ydb_Query.ExecuteQueryRequest, _ ...grpc.CallOption) (
 					Ydb_Query_V1.QueryService_ExecuteQueryClient, error,
 				) {
-					// Simulate session expiry: cancelling ctx starts the AfterFunc goroutine.
+					// Simulate session expiry: canceling ctx starts the AfterFunc goroutine.
 					// stop() called after this returns will always return false.
 					cancel()
 

--- a/internal/query/execute_query_test.go
+++ b/internal/query/execute_query_test.go
@@ -503,24 +503,24 @@ func TestExecute(t *testing.T) {
 		// Regression test for https://github.com/ydb-platform/ydb-go-sdk/issues/2081.
 		//
 		// When the parent ctx is cancelled inside ExecuteQuery (simulating session expiry
-		// while the gRPC stream is still open), the AfterFunc goroutine that propagates
-		// the cancellation to executeCtx is guaranteed to have been started by the time
-		// stop() is called. execute() must detect this via stop() returning false and
-		// return a retryable error — not a non-retryable context.Canceled — so the pool
-		// can retry with a new session.
+		// while the gRPC stream is still open), ctx.Done() is already closed by the time
+		// execute() checks it after ExecuteQuery returns. For idempotent operations,
+		// execute() must detect this via the non-blocking ctx.Done() select and return a
+		// retryable error so the pool can retry with a new session.
 		//
-		// RED before fix: the old code (no stop() check before newResult) would either
-		// let the AfterFunc goroutine race with newResult's Recv() (returning a
-		// non-retryable error) or succeed, making it flaky on multi-core systems.
-		// GREEN after fix: stop() always returns false here (AfterFunc goroutine already
-		// started when cancel() fires ctx.Done()), so execute() deterministically returns
-		// a retryable error without calling Recv() at all.
+		// RED before fix: the old code (no ctx.Done() check before newResult) proceeded
+		// to newResult, which called Recv() on the mock — but no Recv() expectation is
+		// registered, causing a gomock "unexpected call" failure.
+		// GREEN after fix: ctx.Done() is already closed when execute() checks it, so it
+		// returns a retryable error without calling Recv() at all.
 		t.Run("CancelParentContextAfterStreamOpen", func(t *testing.T) {
 			ctrl := gomock.NewController(t)
-			ctx, cancel := context.WithCancel(xtest.Context(t))
+			// Mark the context as idempotent so that cancellation during ExecuteQuery
+			// is classified as a retryable error (non-idempotent returns ctx.Err() directly).
+			ctx, cancel := context.WithCancel(xcontext.WithIdempotent(xtest.Context(t), true))
 
 			// Recv must NOT be called: execute() returns a retryable error before
-			// reaching newResult because stop() detects the AfterFunc was already started.
+			// reaching newResult because ctx.Done() is already closed.
 			stream := NewMockQueryService_ExecuteQueryClient(ctrl)
 
 			client := NewMockQueryServiceClient(ctrl)
@@ -528,8 +528,8 @@ func TestExecute(t *testing.T) {
 				func(_ context.Context, _ *Ydb_Query.ExecuteQueryRequest, _ ...grpc.CallOption) (
 					Ydb_Query_V1.QueryService_ExecuteQueryClient, error,
 				) {
-					// Simulate session expiry: canceling ctx starts the AfterFunc goroutine.
-					// stop() called after this returns will always return false.
+					// Simulate session expiry: canceling ctx closes ctx.Done() so that
+					// the non-blocking check in execute() fires after ExecuteQuery returns.
 					cancel()
 
 					return stream, nil

